### DISCUSSION TITLE
tfvars file

### DISCUSSION THREAD

**Discussion by ramazulay** [D10S0VSkY-OSS/Stack-Lifecycle-Deployment#62] [General]
Hi,
There is a tfvars file in my repository and when I create a deploy(by stack) the field is not completed.
Am I missing something?
Thanks

**Comment by D10S0VSkY-OSS** [comment]
Hello @ramazulay ,
SLD **generates the tfvars dynamically, and the data is passed through the UI via the form or a json through the api rest**.
SLD uses a single stack, respecting the DRY for different deployments and manages the persistence of those variables equally, that's why you can destroy and/or deploy without having to fill out a form again.
**You can download the tfvars , in json format from the api**

**Reply by ramazulay** [reply]
Hi @D10S0VSkY-OSS ,
Thanks for your replay.
I'll be more specific, is there a possibility that SLD will not generate tfvars if existing in repo?

**Reply by D10S0VSkY-OSS** [reply]
Hi @ramazulay 
Not right now, but I can add a feature.  because it seems like a good idea

**Reply by ramazulay** [reply]
Thanks a lot.

**Reply by ramazulay** [reply]
Hi @D10S0VSkY-OSS ,
Regarding the feature above, do you have ETA?

Thanks.

**Reply by D10S0VSkY-OSS** [reply]
Hi @ramazulay 
WIP

**Comment by D10S0VSkY-OSS** [comment]
Hi @ramazulay 
Version v2.4.0 is already available which contains the feature to be able to use different tfvars in a repo.
You only have to indicate the name of the tfvars file in the **tfvar_file** form, obviously it must exist in the repository
Once the tfvars file is indicated, it ignores all types of variables that are passed to it by the form.
regards!

**Reply by ramazulay** [reply]
Thanks a lot.
How do I upgrade my ENV? The SLD runs on K8S in EKS.
I have saw that the docker image in DockerHub was not updated.

Thanks.

**Comment by D10S0VSkY-OSS** [comment]
Hi @ramazulay ,

This change requires the evolution schema of the deploy  table DB , if it is already implemented in production 

1. Export deploy table
2. Drop table deploy
3. Apply kubectl with new versión d10s0vsky/sld-dashboard:2.3.0 and d10s0vsky/sld-api:2.3.0 (when the api-sld pod starts it will create the deployment table in the database with the new schame)
4. Restore export deploy table

if you don't have anything deploy in  production, just destroy the infrastructure you have with SLD, delete all pods and data from the db and recreate it like the first time you did it

**Reply by ramazulay** [reply]
Hi @D10S0VSkY-OSS ,
I'm in production and I've been following your sections.
When I restore the deploy table(mysql -u root -p restapi < deploy.sql), I got this page:
![image](https://user-images.githubusercontent.com/41533452/176443761-b01e4d9f-5036-488f-ba01-9044e75f5a4d.png)

From api-backend log:
"GET /api/v1/deploy?limit=1 HTTP/1.1" 307 Temporary Redirect
"GET /api/v1/deploy/?limit=1 HTTP/1.1" 400 Bad Request
"GET /api/v1/deploy/?limit=15 HTTP/1.1" 400 Bad Request
"GET /api/v1/tasks/id/ HTTP/1.1" 404 Not Found

What I need to do?
BTW - The sld-dashboard:2.3.0 and d10s0vsky/sld-api:2.3.0 upgraded.

**Reply by D10S0VSkY-OSS** [reply]
Hi @ramazulay 
could you pass me output this command:
mysql> desc deploy;

**Reply by ramazulay** [reply]
Hi @D10S0VSkY-OSS ,
![image](https://user-images.githubusercontent.com/41533452/176462470-98c70dec-7d7c-4ad2-a0f5-78e765c81722.png)

**Comment by D10S0VSkY-OSS** [comment]
@ramazulay 
When you restore, you crush the schema. you are missing a field tfvar_file   

```

+--------------+--------------+------+-----+---------+----------------+
| Field        | Type         | Null | Key | Default | Extra          |
+--------------+--------------+------+-----+---------+----------------+
| id           | int          | NO   | PRI | NULL    | auto_increment |
| task_id      | varchar(36)  | YES  |     | NULL    |                |
| name         | varchar(100) | YES  |     | NULL    |                |
| action       | varchar(100) | YES  |     | NULL    |                |
| start_time   | varchar(100) | YES  |     | NULL    |                |
| destroy_time | varchar(100) | YES  |     | NULL    |                |
| stack_name   | varchar(100) | YES  |     | NULL    |                |
| created_at   | datetime     | YES  |     | NULL    |                |
| updated_at   | datetime     | YES  |     | NULL    |                |
| user_id      | int          | YES  |     | NULL    |                |
| username     | varchar(50)  | NO   |     | NULL    |                |
| squad        | varchar(50)  | NO   | MUL | NULL    |                |
| variables    | json         | YES  |     | NULL    |                |
| environment  | varchar(50)  | YES  |     | NULL    |                |
| tfvar_file   | varchar(50)  | YES  |     | NULL    |                |
+--------------+--------------+------+-----+---------+----------------+
```
can you try dump **only data** and restore this

1. mysqldump --skip-triggers --compact --no-create-info -h db -uroot -p123 restapi deploy > sld_deploy.sql
2. drop table deploy
3. restart api-backend pod for recreate table
4. Check if schema have field tfvar_file
5. restore sql dump

**Reply by ramazulay** [reply]
Hi @D10S0VSkY-OSS ,
I have got the error:
ERROR 1136 (21S01): Column count doesn't match value count at row 1

I think because you added a new field tfvar_file, its not possible.

**Reply by D10S0VSkY-OSS** [reply]
@ramazulay 
please, try drop table deploy, restore first dump with schema inlcude  and add field manually like:
```
ALTER TABLE deploy ADD tfvar_file varchar(50) NULL AFTER environment;

**Reply by ramazulay** [reply]
@D10S0VSkY-OSS  Thanks a lot, works well.

**Comment by D10S0VSkY-OSS** [comment]
@ramazulay
cool!
thanks to you for contributing with this great idea, which will allow from the same stack (we respect the dry) to deploy in various environments or accounts defined as IaC

**Comment by ramazulay** [comment]
Hi,
Please follow the files:
1. test.tfvars:
instance_type           = "t2.micro"

2. variables.tf:
variable "instance_type" {
  type = string
  description = "instance type"
  #default = "t2.micro"
}

3. main.tf:
provider "aws" {
  #profile = "default"
  region  = "eu-central-1"
}

resource "null_resource" "python" {
  provisioner "local-exec" {
    command = "python3 --version && pip3 install kubernetes"
  }
}

resource "aws_instance" "test-ram-deploy" {
  ami                         = "ami-XXXXXXX"
  instance_type               = var.instance_type
  count                       = 1
  #ecs_associate_public_ip_address = "false"
  tags = {
    Name        = "test-ram-deploy"
    Environment = "development"
  }
}


output "test-ram-deploy_ips" {
  value = ["${aws_instance.test-ram-deploy.*.public_ip}"]
}

output "test-ram-deploy_dns" {
  value = ["${aws_instance.test-ram-deploy.*.public_dns}"]
}

The "instance_type" does not fill automatically:
[image: image.png]

Thanks.

‫בתאריך יום ג׳, 17 במאי 2022 ב-11:32 מאת ‪Diego La Falce‬‏ <‪
***@***.***‬‏>:‬
